WIP: Add Rasch Example

bdeonovic opened the pull request:
Another example for constrained Optimization. Not sure if I did everything right so its WIP.

anriseth commented:
Great, thank you :)

I'm on my phone so I haven't looked through the actual example file, but here are a few points:

- You'll need to add ForwardDiff to the `docs/Project.toml` file (I think you'll find the UUID in JuliaRegistries)
- I think it'll be good to create a `docs/src/examples/data` folder and then at the bottom of the `docs/generate.jl` file you can add code to copy that whole folder over to the `GENERATEDDIR`
- You need a link to your example, I suggest putting it beneath the likelihood example here https://github.com/JuliaNLSolvers/Optim.jl/blob/master/docs/mkdocs.yml#L38

pkofod commented:
Is this a data set that is in RDatasets.jl or similar?  Or alternatively, could we generate the data? The latter also means that we can test for robustness of the algorithm.

If no and no then I need to be sure that this is OK license-wise

bdeonovic commented:
The dataset is not in RDatasets.jl, but I could generate data that fits the model instead of using this dataset. I'm feeling a bit under the weather today so I'll get to it tomorrow.

pkofod commented:
Could be cool if we could finish this up @bdeonovic !

bdeonovic commented:
Could you check if I have added the example correctly? I took out the data and added simulated data. I'm not sure if I did the markdown stuff right. 

One thing that is weird: I noticed that most of the time the IPNewton works, but sometimes it fails and returns all NaNs. I think this has to do with the way the starting barrier coefficients are chosen. I haven't tested it a lot with the simulated data, but with the data set I had previously it occurred maybe every 10th run or so.

pkofod commented:
> One thing that is weird: I noticed that most of the time the IPNewton works, but sometimes it fails and returns all NaNs. I think this has to do with the way the starting barrier coefficients are chosen. I haven't tested it a lot with the simulated data, but with the data set I had previously it occurred maybe every 10th run or so.

This is *not* good, but thank you so much for reporting it. Have you turned the trace on? I'm wondering if it's initialized as `NaN` or is it turns `NaN` somewhere in the middle. If it's the former it could be finding the first feasible point, if it's the latter I suspect the line search might not guard against non-finite values. It might also be the evaluation of your objective or constraints that might theoretically be finite, but on a computer can become non-finite. Thanks for reporting anyway!

bdeonovic commented:
trace looks like this:
```julia
Iter     Lagrangian value Function value   Gradient norm    |==constr.|      μ
     0   NaN              -Inf             NaN              NaN              1.00e+00
 * c: [0.0]
 * g(x): [NaN, NaN, NaN, NaN, NaN]
 * x: [0.0, 0.0, 0.0, 0.0, 0.0]
 * gtilde(x): [NaN, NaN, NaN, NaN, NaN]
 * h(x): [NaN NaN NaN NaN NaN; NaN NaN NaN NaN NaN; NaN NaN NaN NaN NaN; NaN NaN NaN NaN NaN; NaN NaN NaN NaN NaN]
 * α: 1.0
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [NaN]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [0.0]
     1   NaN              NaN              NaN              NaN              0.00e+00
 * c: [NaN]
 * g(x): [NaN, NaN, NaN, NaN, NaN]
 * x: [NaN, NaN, NaN, NaN, NaN]
 * gtilde(x): [NaN, NaN, NaN, NaN, NaN]
 * h(x): [NaN NaN NaN NaN NaN; NaN NaN NaN NaN NaN; NaN NaN NaN NaN NaN; NaN NaN NaN NaN NaN; NaN NaN NaN NaN NaN]
 * α: 0.0
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [NaN]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [NaN]
  0.140869 seconds (2.47 k allocations: 122.766 KiB)
([NaN, NaN, NaN, NaN], NaN, 1, false, [NaN, NaN, NaN, NaN, NaN])
```

bdeonovic commented:
And if I literally just run it again I get:

```julia
Iter     Lagrangian value Function value   Gradient norm    |==constr.|      μ
     0   1.348190e+03     1.348190e+03     2.108000e+02     0.000000e+00     1.00e+00
 * c: [0.0]
 * g(x): [160.2, -54.8, -210.8, -0.8, 106.2]
 * x: [0.0, 0.0, 0.0, 0.0, 0.0]
 * gtilde(x): [160.2, -54.8, -210.8, -0.8, 106.2]
 * h(x): [137.6 -34.4 -34.4 -34.4 -34.4; -34.4 137.6 -34.4 -34.4 -34.4; -34.4 -34.4 137.6 -34.4 -34.4; -34.4 -34.4 -34.4 137.6 -34.4; -34.4 -34.4 -34.4 -34.4 137.6]
 * α: 1.0
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-5.68434e-14]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [0.0]
     1   1.164769e+03     1.164769e+03     9.953180e+01     7.232640e-16     0.00e+00
 * c: [1.17684e-14]
 * g(x): [85.4339, -32.6196, -99.5318, -5.98103, 53.0058]
 * x: [-0.466055, 0.158945, 0.612433, 0.00196827, -0.307292]
 * gtilde(x): [85.4339, -32.6196, -99.5318, -5.98103, 53.0058]
 * h(x): [108.645 -27.9649 -31.6349 -26.2891 -22.695; -27.9649 141.286 -45.9285 -36.4954 -30.8359; -31.6349 -45.9285 155.057 -42.2878 -35.1443; -26.2891 -36.4954 -42.2878 134.045 -28.9113; -22.695 -30.8359 -35.1443 -28.9113 117.648]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-0.0614583]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-1.17684e-14]
     2   1.093634e+03     1.093634e+03     1.548891e+01     2.244216e-25     0.00e+00
 * c: [-2.357e-13]
 * g(x): [15.4889, -6.04848, -7.48572, -5.39331, 3.4386]
 * x: [-1.06549, 0.410706, 1.16529, 0.105793, -0.616297]
 * gtilde(x): [15.4889, -6.04848, -7.48572, -5.39331, 3.4386]
 * h(x): [68.9618 -18.9198 -19.4248 -17.5681 -13.049; -18.9198 139.47 -53.6619 -40.5237 -26.3648; -19.4248 -53.6619 146.3 -45.431 -27.7823; -17.5681 -40.5237 -45.431 127.683 -24.1607; -13.049 -26.3648 -27.7823 -24.1607 91.3568]
 * α: 1.0
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [9.52148e-13]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [2.357e-13]
     3   1.092145e+03     1.092145e+03     8.017727e+00     3.849665e-16     0.00e+00
 * c: [9.48778e-5]
 * g(x): [8.01763, -3.10331, -3.84491, -2.76244, 1.69304]
 * x: [-1.15354, 0.440898, 1.19851, 0.135298, -0.621074]
 * gtilde(x): [8.01763, -3.10331, -3.84491, -2.76244, 1.69304]
 * h(x): [64.1615 -17.6256 -17.8457 -16.4677 -12.2225; -17.6256 139.282 -54.1839 -41.2029 -26.2696; -17.8457 -54.1839 145.339 -45.9025 -27.4072; -16.4677 -41.2029 -45.9025 127.743 -24.1699; -12.2225 -26.2696 -27.4072 -24.1699 90.0692]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-4.0575e-12]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-9.48778e-5]
     4   1.091723e+03     1.091723e+03     4.090790e+00     6.382055e-16     0.00e+00
 * c: [0.000349079]
 * g(x): [4.09044, -1.5753, -1.95215, -1.40116, 0.838168]
 * x: [-1.20258, 0.457383, 1.21667, 0.151364, -0.622483]
 * gtilde(x): [4.09044, -1.5753, -1.95215, -1.40116, 0.838168]
 * h(x): [61.5783 -16.9229 -17.0074 -15.8639 -11.7841; -16.9229 139.158 -54.4526 -41.5572 -26.2252; -17.0074 -54.4526 144.821 -46.1438 -27.2174; -15.8639 -41.5572 -46.1438 127.745 -24.1801; -11.7841 -26.2252 -27.2174 -24.1801 89.4068]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-1.82825e-12]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-0.000349079]
     5   1.091609e+03     1.091609e+03     2.067984e+00     1.292839e-16     0.00e+00
 * c: [0.000177649]
 * g(x): [2.06781, -0.79419, -0.984177, -0.706133, 0.416693]
 * x: [-1.22875, 0.465984, 1.22615, 0.159734, -0.622946]
 * gtilde(x): [2.06781, -0.79419, -0.984177, -0.706133, 0.416693]
 * h(x): [60.2315 -16.5549 -16.5736 -15.546 -11.557; -16.5549 139.087 -54.5894 -41.7389 -26.2042; -16.5736 -54.5894 144.551 -46.2663 -27.1221; -15.546 -41.7389 -46.2663 127.738 -24.1871; -11.557 -26.2042 -27.1221 -24.1871 89.0703]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [7.27751e-13]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-0.000177649]
     6   1.091580e+03     1.091580e+03     1.039959e+00     1.271615e-16     0.00e+00
 * c: [9.13303e-5]
 * g(x): [1.03987, -0.398822, -0.494211, -0.354538, 0.207704]
 * x: [-1.24228, 0.470403, 1.23102, 0.164032, -0.62309]
 * gtilde(x): [1.03987, -0.398822, -0.494211, -0.354538, 0.207704]
 * h(x): [59.5427 -16.3664 -16.3526 -15.3825 -11.4412; -16.3664 139.05 -54.6585 -41.831 -26.1941; -16.3526 -54.6585 144.413 -46.3281 -27.0743; -15.3825 -41.831 -46.3281 127.733 -24.1911; -11.4412 -26.1941 -27.0743 -24.1911 88.9006]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [1.39233e-12]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-9.13303e-5]
     7   1.091572e+03     1.091572e+03     5.215094e-01     3.722063e-17     0.00e+00
 * c: [4.15412e-5]
 * g(x): [0.521468, -0.199855, -0.247649, -0.177648, 0.103685]
 * x: [-1.24916, 0.472643, 1.23349, 0.16621, -0.62314]
 * gtilde(x): [0.521468, -0.199855, -0.247649, -0.177648, 0.103685]
 * h(x): [59.1942 -16.2709 -16.241 -15.2997 -11.3826; -16.2709 139.031 -54.6932 -41.8774 -26.1891; -16.241 -54.6932 144.344 -46.3591 -27.0504; -15.2997 -41.8774 -46.3591 127.729 -24.1933; -11.3826 -26.1891 -27.0504 -24.1933 88.8154]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [8.95993e-13]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-4.15412e-5]
     8   1.091570e+03     1.091570e+03     2.611514e-01     2.122020e-18     0.00e+00
 * c: [2.83263e-5]
 * g(x): [0.261123, -0.10004, -0.123962, -0.0889205, 0.0518]
 * x: [-1.25263, 0.473773, 1.23474, 0.167309, -0.623157]
 * gtilde(x): [0.261123, -0.10004, -0.123962, -0.0889205, 0.0518]
 * h(x): [59.0189 -16.2228 -16.1849 -15.2579 -11.3532; -16.2228 139.021 -54.7107 -41.9007 -26.1866; -16.1849 -54.7107 144.309 -46.3746 -27.0384; -15.2579 -41.9007 -46.3746 127.728 -24.1944; -11.3532 -26.1866 -27.0384 -24.1944 88.7727]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [7.49134e-14]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-2.83263e-5]
     9   1.091570e+03     1.091570e+03     1.306737e-01     6.797658e-17     0.00e+00
 * c: [1.418e-5]
 * g(x): [0.13066, -0.0500485, -0.0620158, -0.0444845, 0.0258893]
 * x: [-1.25438, 0.474339, 1.23536, 0.167859, -0.623165]
 * gtilde(x): [0.13066, -0.0500485, -0.0620158, -0.0444845, 0.0258893]
 * h(x): [58.931 -16.1987 -16.1568 -15.237 -11.3385; -16.1987 139.016 -54.7194 -41.9123 -26.1854; -16.1568 -54.7194 144.291 -46.3824 -27.0324; -15.237 -41.9123 -46.3824 127.727 -24.195; -11.3385 -26.1854 -27.0324 -24.195 88.7513]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [4.79382e-12]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-1.418e-5]
    10   1.091570e+03     1.091570e+03     6.536313e-02     1.827353e-17     0.00e+00
 * c: [8.76913e-6]
 * g(x): [0.0653544, -0.0250314, -0.0310166, -0.0222483, 0.0129419]
 * x: [-1.25525, 0.474623, 1.23567, 0.168135, -0.623168]
 * gtilde(x): [0.0653544, -0.0250314, -0.0310166, -0.0222483, 0.0129419]
 * h(x): [58.887 -16.1866 -16.1428 -15.2265 -11.3311; -16.1866 139.013 -54.7237 -41.9182 -26.1848; -16.1428 -54.7237 144.282 -46.3863 -27.0294; -15.2265 -41.9182 -46.3863 127.726 -24.1953; -11.3311 -26.1848 -27.0294 -24.1953 88.7406]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [2.08385e-12]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-8.76913e-6]
    11   1.091570e+03     1.091570e+03     3.268777e-02     7.573685e-18     0.00e+00
 * c: [4.43356e-6]
 * g(x): [0.0326833, -0.0125175, -0.0155105, -0.0111257, 0.00647029]
 * x: [-1.25569, 0.474764, 1.23583, 0.168273, -0.62317]
 * gtilde(x): [0.0326833, -0.0125175, -0.0155105, -0.0111257, 0.00647029]
 * h(x): [58.8649 -16.1806 -16.1357 -15.2212 -11.3274; -16.1806 139.012 -54.7259 -41.9211 -26.1845; -16.1357 -54.7259 144.278 -46.3883 -27.0279; -15.2212 -41.9211 -46.3883 127.726 -24.1954; -11.3274 -26.1845 -27.0279 -24.1954 88.7352]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-1.70826e-12]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-4.43356e-6]
    12   1.091570e+03     1.091570e+03     1.634542e-02     6.653951e-18     0.00e+00
 * c: [2.21282e-6]
 * g(x): [0.0163432, -0.00625918, -0.00775578, -0.00556323, 0.00323498]
 * x: [-1.25591, 0.474835, 1.23591, 0.168342, -0.623171]
 * gtilde(x): [0.0163432, -0.00625918, -0.00775578, -0.00556323, 0.00323498]
 * h(x): [58.8649 -16.1806 -16.1357 -15.2212 -11.3274; -16.1806 139.012 -54.7259 -41.9211 -26.1845; -16.1357 -54.7259 144.278 -46.3883 -27.0279; -15.2212 -41.9211 -46.3883 127.726 -24.1954; -11.3274 -26.1845 -27.0279 -24.1954 88.7352]
 * α: 0.5
 * bstate: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-3.007e-12]
 * bgrad: BarrierStateVars{Float64}:
  slack_x: Float64[]
  slack_c: Float64[]
  λx: Float64[]
  λc: Float64[]
  λxE: Float64[]
  λcE: [-2.21282e-6]
  0.337509 seconds (15.42 k allocations: 635.250 KiB)
([-1.25591, 0.474835, 1.23591, 0.168342], -1091.5696924896274, 12, true, [NaN, NaN, NaN, NaN, NaN])
```

pkofod commented:
Right, so it seems like it is indeed one of my guesses: the initial phase fails for some reason, and I guess it's because an array is initialized incorrectly. Thanks !

bdeonovic commented:
Why is the behavior "stochastic" as in it doesn't happen every time?

pkofod commented:
> Why is the behavior "stochastic" as in it doesn't happen every time?

That's what I meant by



> and I guess it's because an array is initialized incorrectly.

pkofod commented:
To clarify: this can happen if you initialize an array as `undef` and then you at some later point (accidentally) use it as if it held zeros in the first iteration.

pkofod commented:
Can you show me exactly what you run to get the NaNs? I mean do you run the script over and over again or just the optimize call, or? I cannot seem to replicate the errors!

pkofod commented:
https://julianlsolvers.github.io/Optim.jl/latest/#examples/generated/rasch/ we may need to fix some latex-stuff in the middle of the text :)
